lib/src/result.dart

```diff
@@ -219,6 +306,8 @@ final class Failure<S, F> extends Result<S, F> {
   @override
   int get hashCode => value.hashCode;
 
+  StackTrace get _stackTrace => StackTrace.current;
```

Comment:
Does StackTrace.current reflect the original source of a problem? Imagine creating a Failure(), passing it between 3 different layers and then trying to consume it. Imho stackTrace should point to where this Failure() was created, not where it was consumed. StackTrace.current will point to consumption place.

Can we capture StackTrace.current in a field when the Failure constructor is called and then reuse it? Then we point specifically to the source

Reply:
`StackTrace.current` is static and available anywhere and is similar to this:
```dart
try { throw 0; } catch (_, stack) { return stack; }
```  
Could you clarify the point of adding it to the contractor? 
ps: you can run or/and modify an example to see how it works now.

Reply:
The point is, why do we have stackTraces? They help us to find a place in the code where the error happened and tell us the history of traces. Consider this example:

```
final operation = await makeCall('x');
if (operation.isFailure) {
  logFailure(operation);
} else {
  // do something with the success
}

Future<Result<String, Exception>> makeCall(String input) {
  if (input.contains('x')) {
    return Failure(Exception('user input is invalid'));
  }
  
  return Success('ok');
}

void logFailure(Result<dynamic, Exception>> result) {
  print(result);
}
```
The above example will print a stackTrace which points to where 'print' is executed inside the `logFailure` method but that's not where the error was generated. What I would like as a developer is to given stackTrace which points to the `return Failure(Exception('user input is invalid'));` inside of `if (input.contains('x')) {`.

Reply:
@dtscalac This is the result of your example execution. Is that what you want to see?
![CleanShot 2024-04-30 at 09 55 36@2x](https://github.com/minikin/result_type/assets/4456735/70c5b34c-f16d-498b-99b1-7fadfec8cbdf)

Reply:
@minikin no, I want to see StackTrace which leads to `makeCall` method, as you see in this output it's not there.

Reply:
@dtscalac, I'd like to see an example of that stack trace. Feel free to propose a solution.

Reply:
Example stackTrace that I want to get:

```
Failure: Exception: user input is invalid, #0      new Failure (package:result_type/src/result.dart:298:49)
#1      makeCall (file:///Users/shemhazai/Downloads/result_type/example/example.dart:24:12)
#2      main (file:///Users/shemhazai/Downloads/result_type/example/example.dart:14:27)
#3      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#4      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

That can be done if we capture the StackTrace when the Failure is created and not when toString() is called. Like that:
```
/// A failure, storing a [Failure] value.
@immutable
final class Failure<S, F extends Exception> extends Result<S, F> {
  final F value;
  final StackTrace stackTrace;

  Failure(this.value) : stackTrace = StackTrace.current;

  @override
  bool operator ==(Object o) {
    if (identical(this, o)) return true;

    return o is Failure<S, F> && o.value == value;
  }

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => 'Failure: $value, $stackTrace';
}
```

I decided to make stacktrace public because many reporting tools support stackTraces as separate fields so that you can call it like:

```
final Failure<Exception> result = await _getResult();
myLogger.logError(
  exception: result.value,
  stackTrace: result.stackTrace,
);
```

Reply:
In my opinion what @dtscalac is proposing is the expected behavior when dealing with the StackTrace: the current behavior "hides" where the real error it's happening while this proposal makes it more explicit.